srt-protocol/src/protocol/mod.rs

```diff
@@ -152,49 +199,85 @@ impl Sub<TimeSpan> for TimeSpan {
     }
 }
 
+impl Add<Duration> for TimeSpan {
+    type Output = TimeSpan;
+
+    fn add(self, rhs: Duration) -> Self::Output {
+        Self(self.0 + rhs.as_micros() as i32)
+    }
+}
+
+impl Sub<Duration> for TimeSpan {
+    type Output = TimeSpan;
+
+    fn sub(self, rhs: Duration) -> Self::Output {
+        Self(self.0 - rhs.as_micros() as i32)
+    }
+}
+
+impl Add<TimeSpan> for Instant {
+    type Output = Instant;
+
+    fn add(self, rhs: TimeSpan) -> Self::Output {
+        let micros = rhs.as_micros() as i64;
+        if micros > 0 {
+            self + Duration::from_micros(micros as u64)
+        } else {
+            self - Duration::from_micros(micros.abs() as u64)
+        }
+    }
+}
+
+impl Sub<TimeSpan> for Instant {
+    type Output = Instant;
+
+    fn sub(self, rhs: TimeSpan) -> Self::Output {
+        let micros = rhs.as_micros() as i64;
+        if micros > 0 {
+            self + Duration::from_micros(micros as u64)
+        } else {
+            self - Duration::from_micros(micros.abs() as u64)
+        }
+    }
+}
+
 impl TimeBase {
     pub fn new(start_time: Instant) -> Self {
-        Self(start_time)
+        Self {
+            origin_time: start_time,
+            reference_time: start_time,
+            reference_ts: TimeStamp::MIN,
+        }
     }
 
     #[allow(clippy::trivially_copy_pass_by_ref)]
     pub fn timestamp_from(&self, instant: Instant) -> TimeStamp {
-        assert!(
-            self.0 <= instant,
-            "Timestamps are only valid after the timebase start time"
-        );
-        TimeStamp(Wrapping(
-            ((instant - self.0).as_micros() & TIMESTAMP_MASK) as u32,
-        ))
+        if instant < self.reference_time {
+            self.reference_ts - (self.reference_time - instant)
+        } else {
+            self.reference_ts + (instant - self.reference_time)
+        }
     }
 
-    // Get Instant closest to `now` that is consistent with `timestamp`
     #[allow(clippy::trivially_copy_pass_by_ref)]
-    pub fn instant_from(&self, now: Instant, timestamp: TimeStamp) -> Instant {
-        let mut wraps = ((now - self.0).as_micros() >> 32) as u64;
-
-        // find the smallest difference
-        let diff = wraps as i64 * WRAP_US as i64 + i64::from(timestamp.as_micros())
-            - (now - self.0).as_micros() as i64;
-        if diff < -i64::from(MAX_DIFF_US) {
-            wraps += 1;
-        } else if diff > i64::from(MAX_DIFF_US) {
-            wraps = wraps.saturating_sub(1);
-        }
-
-        self.0 + Duration::from_micros(wraps * WRAP_US + timestamp.as_micros() as u64)
+    pub fn instant_from(&self, timestamp: TimeStamp) -> Instant {
+        self.reference_time + (timestamp - self.reference_ts)
     }
 
-    pub fn adjust(&mut self, delta: TimeSpan) {
-        if delta.0 > 0 {
-            self.0 += Duration::from_micros(delta.0 as u64);
-        } else {
-            self.0 -= Duration::from_micros(delta.0.abs() as u64);
+    #[allow(clippy::assign_op_pattern)]
```

Comment:
This is fine, maybe more correct to just implement the traits.....

Reply:
Unfortunately I'd have to submit a pull request to the Rust core to do this. The types here are Instant and Duration from std, otherwise I would have certainly implemented the traits as you suggested. :)